JoinColony/block-ingestor
# [Arbitrary Txs] Decode arbitrary function call

[CDapp Issue](https://github.com/JoinColony/colonyCDapp/issues/3841)

This PR adds a util to decode arbitrary function data by trying to look up the signatures via 4byte.directory API and using them to create an ethers interface.

## Testing

1. Copy the FOUR_BYTE_API_URL variable from `.env.example` to `.env`.
1. Apply the patch below to turn `index.ts` into a testing playground (is it time for unit test setup? 🤔). 
2. Run `npm run dev` to verify the test functions are decoded correctly.

You can also run some custom tests.

One option is to use [this encoding tool](https://abi.hashex.org) to encode data for an arbitrary function you specify. Bear in mind the hex hash of this function will have to be known by 4byte.directory.

Alternatively, you can test function data CDapp generates when creating motions. You need to find the `TX DEBUG` log corresponding to `createMotion` call and locate the 4th (index 3) element of the params:
<img width="538" alt="image" src="https://github.com/user-attachments/assets/dca54673-c387-42f5-b7cf-a39347b43f0f">

Add a separate 'test case' to `index.ts` and restart `npm run dev` to check.

Patch (save to a new file, run `git apply <path>`)
```
diff --git a/src/index.ts b/src/index.ts
index 8bc73998..67fa7ab5 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -6,54 +6,76 @@ import 'cross-fetch/polyfill';
 
 import { utils } from 'ethers';
 
-import { startBlockListener } from '~blockListener';
-import amplifyClientSetup from '~amplifyClient';
-import { initialiseProvider } from '~provider';
-import { startStatsServer } from '~stats';
-import {
-  setupListenersForColonies,
-  setupListenersForExtensions,
-} from '~eventListeners';
-import { seedDB } from '~utils';
-import { setupNotificationsClient } from '~utils/notifications';
+// import { startBlockListener } from '~blockListener';
+// import amplifyClientSetup from '~amplifyClient';
+// import { initialiseProvider } from '~provider';
+// import { startStatsServer } from '~stats';
+// import {
+//   setupListenersForColonies,
+//   setupListenersForExtensions,
+// } from '~eventListeners';
+import { decodeArbitraryFunction /* seedDB */ } from '~utils/parseFunction';
+// import { setupNotificationsClient } from '~utils/notifications';
 
 utils.Logger.setLogLevel(utils.Logger.levels.ERROR);
 
 const start = async (): Promise<void> => {
-  amplifyClientSetup();
-
-  /**
-   * Setup the notifications provider so that notifications can be sent when needed
-   */
-  await setupNotificationsClient();
-
-  /**
-   * Start express server providing stats and fetch existing stats from the DB
-   */
-  await startStatsServer();
-
-  /**
-   * Setup the listeners we care about for existing colonies, extensions
-   * This has to be done before the block listener is started to ensure the events are not missed
-   */
-  await setupListenersForColonies();
-  await setupListenersForExtensions();
-
-  /**
-   * Start the main block listener
-   */
-  startBlockListener();
-
-  await initialiseProvider();
-
-  /**
-   * In development, where both the chain and the DB gets reset everytime,
-   * we need to "seed" some initial data, such as versions or the current network fee
-   * In live environments, these values will already have been saved in the DB
-   */
-  if (process.env.NODE_ENV === 'development') {
-    await seedDB();
-  }
+  // makePaymentFundedFromDomain(uint256,uint256,uint256,uint256,address[],address[],uint256[],uint256,uint256)
+  console.log(
+    await decodeArbitraryFunction(
+      '0x2a2678bb0000000000000000000000000000000000000000000000000000000000000001ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000000000000000000000000000000000000000000001ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000000000000000000000000000000000000000000120000000000000000000000000000000000000000000000000000000000000016000000000000000000000000000000000000000000000000000000000000001a0000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000b77d57f4959eafa0339424b83fcfaf9c154074610000000000000000000000000000000000000000000000000000000000000001000000000000000000000000ef841fe1611ce41bfcf0265097efaf50486f51110000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000001a682a318ba28df07c20',
+    ),
+  );
+
+  // mintTokens(uint256)
+  console.log(
+    await decodeArbitraryFunction(
+      '0x97304ced00000000000000000000000000000000000000000000021e0c0013070adc0000',
+    ),
+  );
+
+  // setUserRoles(uint256,uint256,address,uint256,bytes32)
+  console.log(
+    await decodeArbitraryFunction(
+      '0xcd4a72cb0000000000000000000000000000000000000000000000000000000000000001ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000009df24e73f40b2a911eb254a8825103723e13209c0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000006f',
+    ),
+  );
+
+  // emitDomainReputationReward(uint256,address,int256)
+  console.log(
+    await decodeArbitraryFunction(
+      '0x944c650f0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000b77d57f4959eafa0339424b83fcfaf9c154074610000000000000000000000000000000000000000000000000de0b6b3a7640000',
+    ),
+  );
+
+  // amplifyClientSetup();
+  // /**
+  //  * Setup the notifications provider so that notifications can be sent when needed
+  //  */
+  // await setupNotificationsClient();
+  // /**
+  //  * Start express server providing stats and fetch existing stats from the DB
+  //  */
+  // await startStatsServer();
+  // /**
+  //  * Setup the listeners we care about for existing colonies, extensions
+  //  * This has to be done before the block listener is started to ensure the events are not missed
+  //  */
+  // await setupListenersForColonies();
+  // await setupListenersForExtensions();
+  // /**
+  //  * Start the main block listener
+  //  */
+  // startBlockListener();
+  // await initialiseProvider();
+  // /**
+  //  * In development, where both the chain and the DB gets reset everytime,
+  //  * we need to "seed" some initial data, such as versions or the current network fee
+  //  * In live environments, these values will already have been saved in the DB
+  //  */
+  // if (process.env.NODE_ENV === 'development') {
+  //   await seedDB();
+  // }
 };
 
 start();

```


1. Start the CDapp env without block-ingestor (or you can stop the running `ingestor` container in Docker).
4. Copy the `FOUR_BYTE_API_URL` variable from `.env.example` to `.env`.
5.
